frontend/src/firebase/actions.ts

```diff
@@ -366,7 +367,68 @@ export const removeSubTask = (
 
 /*
  * --------------------------------------------------------------------------------
- * Section 3: Other Compound Actions
+ * Section 3: Groups Actions
+ * --------------------------------------------------------------------------------
+ */
+
+export const joinGroup = async (
+  groupId: string,
+): Promise<void> => {
+  const groupDoc = await database.groupsCollection().doc(groupId);
+  const members = await getMembers(groupId);
+  const { email } = getAppUser();
+  // Check if user is already in the group
+  if (members.includes(email)) {
+    return;
+  }
+  groupDoc.update({ members: [...members, email] });
+};
+
+export const createGroup = (
+  name: string,
+  deadline: Date,
+  classCode: string,
+): void => {
+  const { email } = getAppUser();
+  // creator is the only member at first
+  const newGroup: FirestoreGroup = { name, deadline, classCode, members: [email] };
+  database.groupsCollection().doc().set(newGroup);
+};
+
+export const leaveGroup = async (
+  groupId: string,
+): Promise<void> => {
+  const members = await getMembers(groupId);
+  const { email } = getAppUser();
+  const newMembers: string[] = members.filter((m: string) => m !== email);
+  const groupDoc = await database.groupsCollection().doc(groupId);
+  groupDoc.update({ members: newMembers });
+};
+
+export const fetchGroups = async (): Promise<FirestoreGroup[]> => {
+  const groups: FirestoreGroup[] = [];
+  const groupSnapshot = await database.groupsCollection()
+    .where('members', 'array-contains', getAppUser().email)
+    .get();
+  groupSnapshot.forEach((res) => {
+    const g: FirestoreGroup = res.data() as FirestoreGroup;
+    groups.push(g);
+  });
+  return groups;
+};
+
+export const getMembers = async (
+  groupId: string,
+): Promise<string[]> => {
+  const groupDoc = await database.groupsCollection().doc(groupId);
+  const groupSnapshot = await groupDoc.get();
+  const group = await groupSnapshot.data() as FirestoreGroup;
+  return group.members;
+};
```

Comment:
Not sure why do we need to read this from firestore. If the listeners are correct, it should already be in redux store